src/components/structures/auth/ForgotPassword.tsx

```diff
@@ -17,130 +17,98 @@ limitations under the License.
-import ErrorDialog from "../../views/dialogs/ErrorDialog";
 import AuthHeader from "../../views/auth/AuthHeader";
 import AuthBody from "../../views/auth/AuthBody";
 import PassphraseConfirmField from "../../views/auth/PassphraseConfirmField";
-import AccessibleButton from '../../views/elements/AccessibleButton';
 import StyledCheckbox from '../../views/elements/StyledCheckbox';
 import { ValidatedServerConfig } from '../../../utils/ValidatedServerConfig';
+import { Icon as LockIcon } from "../../../../res/img/element-icons/lock.svg";
+import QuestionDialog from '../../views/dialogs/QuestionDialog';
+import { EnterEmail } from './forgot-password/EnterEmail';
+import { CheckEmail } from './forgot-password/CheckEmail';
+import Field from '../../views/elements/Field';
+import { ErrorMessage } from '../ErrorMessage';
+import { Icon as CheckboxIcon } from "../../../../res/img/element-icons/Checkbox.svg";
+import { VerifyEmailModal } from './forgot-password/VerifyEmailModal';
+import Spinner from '../../views/elements/Spinner';
+import { formatSeconds } from '../../../DateUtils';
 
 enum Phase {
-    // Show the forgot password inputs
-    Forgot = 1,
+    // Show email input
+    EnterEmail = 1,
     // Email is in the process of being sent
     SendingEmail = 2,
     // Email has been sent
     EmailSent = 3,
-    // User has clicked the link in email and completed reset
-    Done = 4,
+    // Show new password input
+    PasswordInput = 4,
+    // Password is in the process of being reset
+    ResettingPassword = 5,
+    // All done
+    Done = 6,
 }
 
-interface IProps {
+interface Props {
     serverConfig: ValidatedServerConfig;
     onServerConfigChange: (serverConfig: ValidatedServerConfig) => void;
     onLoginClick?: () => void;
     onComplete: () => void;
 }
 
-interface IState {
+interface State {
     phase: Phase;
     email: string;
     password: string;
     password2: string;
-    errorText: string;
-
-    // We perform liveliness checks later, but for now suppress the errors.
-    // We also track the server dead errors independently of the regular errors so
-    // that we can render it differently, and override any other error the user may
-    // be seeing.
-    serverIsAlive: boolean;
-    serverErrorIsFatal: boolean;
-    serverDeadError: string;
-
-    currentHttpRequest?: Promise<any>;
+    errorText: string | null;
 
     serverSupportsControlOfDevicesLogout: boolean;
     logoutDevices: boolean;
 }
 
-enum ForgotPasswordField {
-    Email = 'field_email',
-    Password = 'field_password',
-    PasswordConfirm = 'field_password_confirm',
-}
-
-export default class ForgotPassword extends React.Component<IProps, IState> {
+export default class ForgotPassword extends React.Component<Props, State> {
     private reset: PasswordReset;
+    private fieldPassword: Field | null = null;
+    private fieldPasswordConfirm: Field | null = null;
 
-    state: IState = {
-        phase: Phase.Forgot,
+    state: State = {
```

Comment:
we should be setting this in the constructor, if we're in the area doing refactorings.

Reply:
Done